src/ace/managers/game.c

```diff
@@ -5,108 +5,128 @@
+		pGameManager->pState->cbSuspend();
 	}
 
-	g_sGameManager.ubStateCount -= 1;
-	g_sGameManager.pStateFirst = pGameState->pPrev;
+	pGameState->pStatePrev = pGameManager->pState;
+	pGameManager->pState = pGameState;
 
-	memFree(pGameState, sizeof(tGameState));
+	if (pGameManager->pState && pGameManager->pState->cbCreate) {
+		pGameManager->pState->cbCreate();
+	}
+
+	logBlockEnd("gameStatePush()");
 }
 
-void gameChangeState(tGameCb cbCreate, tGameCb cbLoop, tGameCb cbDestroy) {
-	logBlockBegin(
-		"gameChangeState(cbCreate: %p, cbLoop: %p, cbDestroy: %p)",
-		cbCreate, cbLoop, cbDestroy
-	);
-	if (g_sGameManager.pStateFirst->cbDestroy) {
-		g_sGameManager.pStateFirst->cbDestroy();
-	}
-	g_sGameManager.pStateFirst->cbCreate = cbCreate;
-	g_sGameManager.pStateFirst->cbLoop = cbLoop;
-	g_sGameManager.pStateFirst->cbDestroy = cbDestroy;
-	g_sGameManager.isRunning = 1;
+void gameStatePop(tGameManager *pGameManager) {
+	logBlockBegin("gameStatePop(pGameManager: %p)", pGameManager);
 
-	if (g_sGameManager.pStateFirst->cbCreate) {
-		g_sGameManager.pStateFirst->cbCreate();
+	if (pGameManager->pState && pGameManager->pState->cbDestroy) {
+		pGameManager->pState->cbDestroy();
 	}
-	logBlockEnd("gameChangeState");
-}
 
-void gameChangeLoop(tGameCb cbLoop) {
-	g_sGameManager.pStateFirst->cbLoop = cbLoop;
-}
+	tGameState *pGameStateToPop = pGameManager->pState;
+	pGameManager->pState = pGameStateToPop->pStatePrev;
 
-void gameProcess(void) {
-	if (g_sGameManager.pStateFirst->cbLoop) {
-		g_sGameManager.pStateFirst->cbLoop();
+	if (pGameManager->pState && pGameManager->pState->cbResume) {
+		pGameManager->pState->cbResume();
 	}
-}
 
-void gameClose(void) {
-	g_sGameManager.isRunning = 0;
+	logBlockEnd("gameStatePop()");
 }
 
-void gameDestroy(void) {
-	tGameState *pGameState, *pPrev;
+void gameStatePopAll(tGameManager *pGameManager) {
+	logBlockBegin("gameStatePopAll(pGameManager: %p)", pGameManager);
 
-	pGameState = g_sGameManager.pStateFirst;
-	while (pGameState) {
-		if (pGameState->cbDestroy) {
-			pGameState->cbDestroy();
+	while (pGameManager->pState) {
+		if (pGameManager->pState->cbDestroy) {
+			pGameManager->pState->cbDestroy();
 		}
-		pPrev = pGameState->pPrev;
-		memFree(pGameState, sizeof(tGameState));
-		pGameState = pPrev;
+
+		pGameManager->pState = pGameManager->pState->pStatePrev;
 	}
+
+	logBlockEnd("gameStatePopAll()");
 }
 
-void gameKill(char *szError) {
-	gameDestroy();
-	systemKill(szError);
+void gameStateChange(tGameManager *pGameManager, tGameState *pGameState) {
+	logBlockBegin("gameStateChange(pGameManager: %p, pGameState: %p)", pGameManager, pGameState);
+
+	if (pGameManager->pState && pGameManager->pState->cbDestroy) {
+		pGameManager->pState->cbDestroy();
+	}
+
+	pGameManager->pState = pGameState;
+
+	if (pGameManager->pState && pGameManager->pState->cbCreate) {
+		pGameManager->pState->cbCreate();
+	}
+
+	logBlockEnd("gameStateChange()");
 }
+
+void gameStateProcess(tGameManager *pGameManager) {
```

Comment:
add `const` to arg, I think

Reply:
At first, I had `const`s, but unfortunately compiler was complaining a lot without type casting.